Add CPU model support to target platform

RossComputerGuy opened the pull request:
## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

Adds a `cpuModel` attribute to `targetPlatform` and defaults the CPU model to `"generic"`. Might need a lot of work but I think this is a great start. We might want to check the CPU model based on the CPU arch coming in and have a fallback of only `"generic"` when we cannot match a recognized CPU arch.

Fixes #313530

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

RossComputerGuy committed:
{lib.systems.parse,stdenv.cc}: add cpu model

Implements the "cpuModel" attribute into system and pass it into the CC
wrapper. This is meant as a multi-tool alternative to "gcc.cpu".

Ericson2314 commented:
This seems close to https://github.com/NixOS/nixpkgs/blob/master/lib/systems/architectures.nix ? My thought at the time with that is that I don't really know what the difference between an instruction set architecture and a "sub instruction set architecture", except for rather arbitrary conventions about what goes in the target triple or not, which I view as sort of historical accident we shouldn't take too seriously.

RossComputerGuy commented:
> This seems close to https://github.com/NixOS/nixpkgs/blob/master/lib/systems/architectures.nix ? My thought at the time with that is that I don't really know what the difference between an instruction set architecture and a "sub instruction set architecture", except for rather arbitrary conventions about what goes in the target triple or not, which I view as sort of historical accident we shouldn't take too seriously.

Yeah, though this mechanism is simpler. I'm not sure how that method works in Nixpkgs but this way you just add your CPU model to the end of config. It also allows for easy checking of common CPU models so marking certain packages as broken can be simpler. Adding this into the config stuff also allows for easy cross compiling via `systems.examples`.

RossComputerGuy commented:
Idk what is up with `by-name` failing. I cannot reproduce that error with the information given.

RossComputerGuy commented:
Just cleaned up the commits, fixed some of the issues found with CI. Made it so this doesn't rebuild everything if the CPU model is generic. I think this is at a state that I am happy.

RossComputerGuy commented:
@ofborg build pkgsCross.asahi-m1.bash

alyssais commented:
> > This seems close to https://github.com/NixOS/nixpkgs/blob/master/lib/systems/architectures.nix ? My thought at the time with that is that I don't really know what the difference between an instruction set architecture and a "sub instruction set architecture", except for rather arbitrary conventions about what goes in the target triple or not, which I view as sort of historical accident we shouldn't take too seriously.
> 
> Yeah, though this mechanism is simpler.

Does this allow us to get rid of architectures.nix?  It's really annoying to have to extend that list with boilerplate when adding new architectures.

Ericson2314 commented:
> Does this allow us to get rid of `architectures.nix`? It's really annoying to have to extend that list with boilerplate when adding new architectures.

Yeah bottom line is I want a unified way of doing things. If we are doing this, then I think it's a hard requirement that is replaces `architecture.nix`. Having multiple ways of specifying things was exactly the sort of thing that the big `lib/systems` reorg I did was supposed to solve --- I don't want it creeping back.

RossComputerGuy commented:
> > > This seems close to https://github.com/NixOS/nixpkgs/blob/master/lib/systems/architectures.nix ? My thought at the time with that is that I don't really know what the difference between an instruction set architecture and a "sub instruction set architecture", except for rather arbitrary conventions about what goes in the target triple or not, which I view as sort of historical accident we shouldn't take too seriously.
> > 
> > 
> > Yeah, though this mechanism is simpler.
> 
> Does this allow us to get rid of architectures.nix? It's really annoying to have to extend that list with boilerplate when adding new architectures.

Yes. This is a simpler alternative so we probably could deprecate it and remove it. This method only requires a couple lines to be changed and models which aren't added to the list could just have all the info specified as an attribute set instead of a string.

Ericson2314 commented on pkgs/build-support/cc-wrapper/default.nix:
```diff
@@ -634,6 +634,12 @@ stdenvNoCC.mkDerivation {
     + optionalString ((targetPlatform ? gcc.cpu) && (isClang || !(targetPlatform.isDarwin && targetPlatform.isAarch64))) ''
       echo "-mcpu=${targetPlatform.gcc.cpu}" >> $out/nix-support/cc-cflags-before
```
Yes it should replace it too, ideally.

RossComputerGuy replied:
Yes, I just didn't strip out the way things are done with `gcc.tune` because I didn't want to break too much. My initial goal was to just get this mechanism to work.

Ericson2314 replied:
Unlike getting rid of `architecture.nix`, I am fine saving that for later :).

RossComputerGuy replied:
Yeah, that was my plan. Maybe add a couple warnings that the file is up for removal so that we can give people a buffer

alyssais commented on pkgs/development/compilers/zig/generic.nix:
```diff
@@ -64,6 +65,9 @@ stdenv.mkDerivation (finalAttrs: {
   meta = {
+    # Zig does not support Apple Silicon on Linux yet.
+    # Ref: https://github.com/ziglang/zig/issues/16331
+    broken = stdenv.isLinux && targetPlatform.isAppleSilicon;
```
This is a conflation I don't think we should be making — whether Zig is broken doesn't depend on whether we're specifically building for Apple Silicon.  The package is equally broken on Apple Silicon when built with the generic CPU model.

RossComputerGuy replied:
That is true but I don't see how we could determine the package is broken and the model is generic but being built on Apple Silicon. I mostly did it as a placeholder for writing a patch that makes Zig working on Apple Silicon and to test the `isAppleSilicon` attribute.

alyssais replied:
But such a patch needs to be applied for all aarch64-linux — we can't force users to specify their CPU and rebuild the world just to have programs that work on their CPU.

alyssais replied:
Can't we just always build Zig on aarch64 for a target with 64K pages, then?

RossComputerGuy replied:
Maybe but who knows what other issues that could bring

RossComputerGuy replied:
Yeah though this feels like it kinda gets off topic on this particular PR. Could raise an issue with the maintainers and get their thoughts.

alyssais replied:
I agree, but I'm bringing it up because I don't think that isAppleSilicon has really been justified.  Simplifying specifying CPU-specific compiler flags seems good, but I don't think it's ever really going to be the right thing to have conditionals based on them.

RossComputerGuy replied:
So should we remove it from this PR or just keep it as is until a follow up PR is made to actually fix Zig?

alyssais replied:
We should remove it from this PR.  There should be no predicates for CPU type.

(Another reason I thought of overnight: Apple Silicon doesn't even necessarily imply 16K pages! You can run a 4K kernel, although lots of stuff currently won't work, it's not some fundamental limitation of the CPU.)

RossComputerGuy replied:
Done

alyssais replied:
No it isn't?  The CPU type predicates are still there.

RossComputerGuy replied:
Where? I don't see it in the Zig files.

alyssais replied:
Ah, when I say that "there should be no predicates for CPU type", I meant that they shouldn't exist, not just that they shouldn't be used in the Zig package.  As I said, I would be surprised if having conditionals based on target CPU (as opposed to passing it to a compiler, etc.) is ever the right thing to do.

RossComputerGuy replied:
Alright, done.

RossComputerGuy commented:
Tried getting ofborg to build bash for Apple Silicon M1 Linux but it seems that bash didn't like the compiler:
```
checking for aarch64-unknown-linux-gnu-gcc... aarch64-unknown-linux-gnu-gcc
checking whether the C compiler works... no
configure: error: in `/build/bash-5.2':
configure: error: C compiler cannot create executables
See `config.log' for more details
```

RossComputerGuy commented:
@ofborg build pkgsCross.asahi-m1.pkgsLLVM.bash

RossComputerGuy commented:
Interesting quick look at the optimizations this can produce. In this example, I used bash.

Standard - 1.1MB
LLVM - 971KB
LLVM w/ CPU model optimizations - 970KB

We save about 1KB of memory on Apple Silicon for bash.

RossComputerGuy committed:
lib.systems: remove architecture

alyssais commented on lib/systems/parse.nix:
```diff
@@ -421,6 +421,61 @@ rec {
+        llvm = "cortex-a53";
+        gnu = "cortex-a53";
+        zig = "cortex_a53";
+      };
```
So far all of these follow the same pattern, of being the same apart from underscores vs hyphens.  Are there counterexamples?

RossComputerGuy replied:
That's true though in the case of Apple Silicon, the model isn't supported by GCC yet. Should I make a simple wrapper which generates the names from a single name?

RossComputerGuy replied:
Might be able to get rid of the `zig` one and just replace `-` with `_` at least.

RossComputerGuy replied:
I made some adjustments so now the function to make the CPU model definition simply can be used for most systems. I decided to keep `zig` in because of `baseline`.

yunfachi commented on lib/systems/default.nix:
```diff
@@ -82,6 +86,13 @@ let
       # Derived meta-data
       useLLVM = final.isFreeBSD;
+      cpuModel = let
+        name = if args ? cpuModel then args.cpuModel else "generic";
```
```suggestion
        name = args.cpuModel or "generic";
```
can be simplified with `or`

RossComputerGuy replied:
Done

RossComputerGuy committed:
zig: use cpu model from target platform

RossComputerGuy commented:
Rebased due to conflicts.

RossComputerGuy commented:
Going to rebase this tonight to work with Zig 0.13 since that PR was merged yesterday. Gentle ping to see what progress we can get.

alyssais commented:
> Gentle ping to see what progress we can get.

AFAICT this still hasn't been addressed:

> If we are doing this, then I think it's a hard requirement that is replaces `architecture.nix`.

RossComputerGuy commented:
@alyssais My understanding with that was to deprecate it here and removed it in a follow up PR.

alyssais commented:
I don't want to see us with _three_ different ways to configure this — two is already too many, and if we put this in without getting rid of (or at least properly deprecating, with a plan to remove — e.g. by implementing the old interface on top of the new one), there's no guarantee we won't end up stuck with three.

(Personally I feel it'd be much easier to get rid of the `gcc` system options than architecture.nix, because they could easily be reimplemented on top of this, but I don't really care which one goes, as long as we don't end up with three.)

RossComputerGuy commented:
Alright, I can try working on stripping the `architectures.nix` out tonight or whenever I have time next.

Ericson2314 commented:
Yes, removing `architectures.nix` is a requirement for me too

RossComputerGuy commented:
@Ericson2314 Should that be included with this PR or a separate one?

RossComputerGuy committed:
openssl: replace gcc.arch with cpu model

RossComputerGuy committed:
libsForQt5.qt5.qtwebengine: replace gcc.arch with cpu model

RossComputerGuy commented:
I've replaced what I can find using `gcc.arch` and changed a few things around. `lib.systems.architecture` throws an error and I think we can remove it after the next NixOS release. The error message is in place to let people know about the change.

RossComputerGuy committed:
nixos/nix: replace gcc.arch with cpu model

reckenrode commented on lib/systems/platforms.nix:
```diff
@@ -385,16 +403,15 @@ rec {
-      arch = "armv8.3-a+crypto+sha2+aes+crc+fp16+lse+simd+ras+rdm+rcpc";
-      cpu = "apple-a13";
-    };
+    cpuModel = "apple_m1";
```
This should fix https://github.com/NixOS/nixpkgs/issues/166401? The reason for the lengthy arch string was due to LLVM 11, which isn’t even in nixpkgs anymore.

RossComputerGuy replied:
Likely, needs testing though.

RossComputerGuy commented:
Just added the ability to detect which compiler versions are being used and adjust some stuff accordingly.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/4278

RossComputerGuy commented:
Will fix conflicts tonight

RossComputerGuy committed:
openblas: remove cpu model when using dynamic arch

RossComputerGuy committed:
openblas: fix formatting

AndersonTorres commented on lib/systems/default.nix:
```diff
@@ -25,7 +25,24 @@ let
   examples = import ./examples.nix { inherit lib; };
-  architectures = import ./architectures.nix { inherit lib; };
+
+  architectures = throw "lib.systems.architectures was deprecated and replaced via the new cpuModel attribute";
```
Deprecated has a specific meaning of "still works but we will remove in the future".
```suggestion
  architectures = throw "architectures was replaced by `cpuModel` attribute";
```

RossComputerGuy replied:
Alright, changed.

RossComputerGuy commented:
Closing in favor of #376197
